### PR TITLE
Add flag to add theme_dir from command line

### PR DESCRIPTION
**Pull request by pfultz2**
This adds a flag to specify the `theme_dir` from the command line.

### CODE REVIEW

**Comment by waylan** [comment]
Just curios, what is the use case for this? Seems to me that you would always want the theme dir to be controlled by the config file. But maybe I'm missing something.

**Comment by pfultz2** [comment]
> Just curios, what is the use case for this?

Its the same use case for the `--theme` flag. You want to be able to specify a different theme at build time. Since, I am using a custom theme, then it needs to be specified using `theme_dir` instead of just `theme`.

**Comment by waylan** [comment]
I understand using a different theme at build time (with the `--theme` flag. However, under what situation would there be more than one `theme_dir` outside of the  `--theme` flag.

To clarify, as I understand it, a complete theme should be in a separate package and called with the  `--theme` flag. The `theme_dir` would be where the user would provide a few replacements/overrides/additions to the theme defined in  `--theme`. The only reason to override `theme_dir` at build time is if you have an entire interdependent theme there, which is not the intended purpose. In fact, I would recommend against it as the `theme` will still be loaded, with `theme_dir` overriding it. If you are not careful (or in future updates), you could get some remnants of `theme` still coming through.

So my point is that if it is bad practice to have an entire theme in `theme_dir`, then a `--theme_dir` flag should not be supported as a means to discourage it. Granted, this would be more clear if #607 was completed.

**Comment by waylan** [comment]
Hmm, although perhaps if `theme=None`, then maybe `theme_dir` could be an independent theme. Not sure if that works though.

**Comment by pfultz2** [comment]
> To clarify, as I understand it, a complete theme should be in a separate package and called with the --theme flag.

Exactly how is that accomplished with a custom theme? Where should custom themes be installed so they are found by the `--theme`? Because right now it looks like `--theme` only gives the ability to use a set prescribed themes.(ie `flatly|mkdocs|simplex|cosmo|readable|slate|amelia|cyborg|readthedocs|spacelab|united|bootstrap|cerulean|journal|yeti`)

**Comment by pfultz2** [comment]
> Hmm, although perhaps if theme=None, then maybe theme_dir could be an independent theme. Not sure if that works though.

Thats exactly how I use `theme_dir`.

**Comment by waylan** [comment]
> Where should custom themes be installed so they are found by the `--theme`? 

I don't believe it is documented, but the themes use setuptools entrypoints. When you install a theme as a separate Python package, it gets registered as a MkDocs theme and becomes available by the name it was registered as. In fact, everything except the `mkdocs` and `readthedocs` themes are separate packages which use this method (see [here][1] and [here][2]). The difference with them is that they are listed as dependencies and get installed by default when you install MkDocs. However, that may change in the future. And there are other [third-party themes][3] which you can install yourself.

[1]: https://github.com/mkdocs/mkdocs-bootswatch
[2]: https://github.com/mkdocs/mkdocs-bootstrap
[3]: https://github.com/mkdocs/mkdocs/wiki/MkDocs-Themes

**Comment by waylan** [comment]
> > Hmm, although perhaps if theme=None, then maybe theme_dir could be an independent theme. Not sure if that works though.

> Thats exactly how I use theme_dir.

Okay, that is the use case I was looking for to justify this. I'm still not sure its a good idea, but I can at least understand it.

**Comment by d0ugal** [comment]
> I don't believe it is documented, but the themes use setuptools entrypoints. When you install a theme as a separate Python package, it gets registered as a MkDocs theme and becomes available by the name it was registered as.

It is documented! http://www.mkdocs.org/user-guide/custom-themes/#packaging-themes

Granted the documentation may not be great. I'd love feedback suggestions (or maybe we just need to make it more visible somehow?)

**Comment by d0ugal** [comment]
> Hmm, although perhaps if theme=None, then maybe theme_dir could be an independent theme. Not sure if that works though.

It does work and it is most useful when you are working on a theme itself. For example, this is how the official packaged themes use their own theme - https://github.com/mkdocs/mkdocs-bootswatch/blob/master/mkdocs.yml#L10

It means you don't need to worry about installing dev versions or maybe I am already using a MkDocs install that has the theme included.

This use case is why I could see this addition being useful. I want to quickly test the MkDocs documentation (or any other docs) against a theme that I am working on.

File: mkdocs/__main__.py [hunk 1]
```diff
@@ -99,9 +100,10 @@ def cli():
 @click.option('-a', '--dev-addr', help=dev_addr_help, metavar='<IP:PORT>')
 @click.option('-s', '--strict', is_flag=True, help=strict_help)
 @click.option('-t', '--theme', type=click.Choice(theme_choices), help=theme_help)
+@click.option('-e', '--theme-dir', type=click.Path(), help=theme_dir_help)
```
Comment (d0ugal) [inline mkdocs/__main__.py]: Does `-e` make sense? `-d` seems more logical to me.

Comment (pfultz2) [inline mkdocs/__main__.py]: `-d` is already used for `--site-dir`. I didn't want to have different options between `serve` and `build`.

**Comment by d0ugal** [comment]
> > > Hmm, although perhaps if `theme=None`, then maybe `theme_dir` could be an independent theme. Not sure if that works though.

> > Thats exactly how I use theme_dir.

> Okay, that is the use case I was looking for to justify this. I'm still not sure its a good idea, but I can at least understand it.

I think it is a reasonable use case. For example, a theme with company branding doesn't belong on PyPI. Then if it isn't going to be distributed, the effort required for packaging is higher than it is worth (particularly if you are not familiar with the lovely quirks of Python packaging.)

So, until we revisit theme customisation etc. more generally, I think we should make it easier to use the current options. I'd be happy for this change to be added. For that to happen, we need an addition in the release notes (under version 0.16.0).

Otherwise the change looks good to me, but I'd like to give it a quick test at some point too.

File: mkdocs/__main__.py (hunk 1 above)
Comment (waylan) [inline mkdocs/__main__.py]: `-d` is used for [`--site_dir`][1] on the build command. And as `--theme_dir` is being added to both `build` and `serve`, the short flag should be consistent (`--site_dir` makes no sense for `serve`).

[1]: https://github.com/mkdocs/mkdocs/blob/master/mkdocs/__main__.py#L127

Comment (d0ugal) [inline mkdocs/__main__.py]: So it is! Sorry, I forgot. I think this is probably the best option then.

**Comment by pfultz2** [comment]
Ok, so I added a note about the new flag under the release notes for version 0.16.

**Comment by d0ugal** [comment]
Thanks!

**Pull request merged**